split up sbatch commands

cjkeilig opened the pull request:
Find SBATCH lines in the commands editor and insert them accordingly, preset the editor with an example if the job is a new one.

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -40,6 +40,10 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
     });
+    editor.setValue("#SBATCH --option=value\n\n# Commands\n\necho \"hello\"");
+    editor.on("focus",function() {
+      editor.setValue("");
```
Wouldn't this clear it each time the user clicks there?

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -124,21 +128,40 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
+    job.commands = editor.getValue();
+    var other = editor.getValue().split("\n");
+    var sbatch = [];
+    var other = angular.forEach(other, function(line, index) {
```
Should probably just use a regular javascript for each loop:

```javascript
for (variable of iterable) {
  statement
}
```

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -124,21 +128,40 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
+    var other = angular.forEach(other, function(line, index) {
+      if (line.includes("SBATCH")) {
+        this.push(line);
+        other.splice(index,1);
```
Which `other` is this splicing?  There is the `var other` that you set this equal to, and the `var other` above that is the contents of the editor.

cjkeilig commented:
1. I do not know why the commands are being saved in job history, I don't know who did that.
2. The placeholder is now regular text
3. I did a traditional for loop, didn't even notice or consider `for (x of y)`, I can rework with 'of'
4. Removed 1 of the "other" variables

5. The "no full path" error is also addressed with the directive at the bottom, for now, nothing happens other than the notice appearing below each path input field. Wanted your thoughts, it would be easier to pattern the path for '/work/<group>/<username>/file.slurm', but I know people may need special cases.

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -124,21 +129,38 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
+    var other = editor.getValue().split("\n");
+    var sbatch = [];
+    for(var x = 0; x < other.length; x++) {
+      if (other[x].includes("SBATCH")) {
```
Make sure it starts with #SBATCH, not just contains.

djw8605 commented on HCCGo/app/js/ConnectionService.js:
```diff
@@ -358,6 +358,23 @@ connectionModule.factory('connectionService',['$log', '$q', '$routeParams', '$lo
+         });
+     }
+   }*/
+
```
Delete all these lines.

djw8605 commented on HCCGo/app/js/ConnectionService.js:
```diff
@@ -891,6 +908,7 @@ connectionModule.factory('connectionService',['$log', '$q', '$routeParams', '$lo
    checkWritable: checkWritable,
    getFileText: getFileText,
    getFileSize: getFileSize,
+   //changeDir: changeDir,
```
Delete

djw8605 commented:
> I do not know why the commands are being saved in job history, I don't know who did that.

I misunderstood how you were implementing this.  All is good.

cjkeilig commented:
1. Made sure to check 'starts with #SBATCH'
2. Removed unneeded method in ConnectionService
3. Deleted line

4. Reworked the directive into an ng-change for the error, catching this no path error as a $validator function didn't work as I did not want to allow invalid input.

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -55,6 +60,14 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
+    if($scope.job.location.search(/^\w*(\.slurm|\.err|\.out)$/)!=-1 |
+     $scope.job.error.search(/^\w*(\.slurm|\.err|\.out)$/)!=-1 | 
+     $scope.job.output.search(/^\w*(\.slurm|\.err|\.out)$/)!=-1){
+     $scope.jobscript.changePath = true;
```
So, if any of the location, error, output has a wrong page, then all of those spans show up?

cjkeilig replied:
Yes

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -290,3 +330,22 @@ jobSubmissionModule.directive('remoteWritable', function($q, $log, connectionSer
+      }
+    }
+  }
+});*/
```
Delete this?

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -124,21 +137,48 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
+      })
+    }
+    else {
+      getWorkProm.when();
```
I don't understand this `when` placement.

cjkeilig replied:
I thought this resolved the promise if it wasn't already

cjkeilig commented:
1. The individual ng-change functions set the path error. The path error works with the other errors.
2. Fixed the when() promise confusion

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -55,6 +60,39 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
+        }
+      }
+  }
+
```
There's gotta be a better way than doing the same function 3 times.

djw8605 replied:
What if you make 1 function.  Pass in the variable and something else.  So the `ng-change` for the error file would look like:

    ... ng-change="chkDir(job.error, 'error')"

Then, then function would look something like:

```javascript
$scope.chkDir = function(dir, identifier) {
  if (dir) {
    if (dir.search(/^\w*\.\w*$/)!=-1) {
      $scope.job[identifier]["changedir"] = true;
    }
    else {
      $scope.job[identifier]["changedir"] = false;
    }
  }
}
```

Then, the `span`s would look like:

```html
 ... ng-show="job.error.changedir" ...
```

Does this make sense?

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -124,21 +143,43 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
+    for(var x = 0; x < other.length; x++) {
+      if (other[x].startsWith("#SBATCH")) {
+        sbatch.push(other[x]);
+        other.splice(x,1);
```
Since you're splicing, you can't iterate over an array without doing something with x.  IE, you are removing an element, so everything shifts down 1 place, but X is still increasing.  So you are missing lines.

cjkeilig commented:
yikes, nice catch. I also fixed up the default editor text. We can put default lines for ntasks/nodes if you would like

djw8605 commented:
Lets do that in a different pull request.

djw8605 commented on HCCGo/app/js/jobSubmissionCtrl.js:
```diff
@@ -40,6 +40,11 @@ jobSubmissionModule.controller('jobSubmissionCtrl', ['$scope', '$log', '$timeout
     });
+
+    // Put a placeholder into the commands editor
+    editor.setValue("#SBATCH --option=\"value\"\n\n# Commands\n\necho \"Hello\"");
```
Where is this value cleared?

cjkeilig replied:
It is not cleared, they just have it there until they change it.
